File: app-ios-nativa/Fontaneria/Fuentes/Fontaneria/ClienteApi.swift
import Dominio
import Foundation

/// Lo que responde el servidor al canjear un código o al renovar la sesión.
public struct RespuestaCanje: Decodable, Sendable {
    public let tokenAcceso: String
    public let expiraEn: MarcaDeTiempo
    public let tokenRefresco: String
    public let usuario: UsuarioApi?
}

public struct UsuarioApi: Decodable, Sendable, Equatable {
    public let id: Int
    public let email: String
    public let proveedor: String
}

public struct RespuestaMetadatos: Decodable, Sendable, Equatable {
    public let titulo: String?
    public let descripcion: String?
    public let imagenUrl: String?
    public let tipo: String

    public var comoMetadatos: MetadatosExtraidos {
        MetadatosExtraidos(
            titulo: titulo,
            descripcion: descripcion,
            imagenUrl: imagenUrl,
            tipo: TipoElemento(rawValue: tipo) ?? .enlace
        )
    }
}

/// Una entrada que el servidor no ha aplicado. Hay que sacarla de la cola
/// igualmente: no se va a aceptar por mucho que se insista, y dejarla dentro
/// reenvía el lote entero en cada sincronización, para siempre.
public struct EntradaRechazada: Decodable, Sendable, Equatable {
    public let id: String
    /// `sin_url`, `sin_nombre` o `no_aplicable` (ver CONTRATO-API.md).
    public let motivo: String
}

public struct RespuestaPush: Decodable, Sendable {
    enum CodingKeys: String, CodingKey {
        case elementos, rechazados, etiquetasDefinidas, etiquetasRechazadas
    }

    public let elementos: [Elemento]
    public let rechazados: [EntradaRechazada]
    public let etiquetasDefinidas: [EtiquetaDefinida]
    public let etiquetasRechazadas: [EntradaRechazada]

    public init(from decoder: any Decoder) throws {
        let campos = try decoder.container(keyedBy: CodingKeys.self)
        elementos = try campos.decodeIfPresent([Elemento].self, forKey: .elementos) ?? []
        rechazados = try campos.decodeIfPresent([EntradaRechazada].self, forKey: .rechazados) ?? []
        etiquetasDefinidas =
            try campos.decodeIfPresent([EtiquetaDefinida].self, forKey: .etiquetasDefinidas) ?? []
        etiquetasRechazadas =
            try campos.decodeIfPresent([EntradaRechazada].self, forKey: .etiquetasRechazadas) ?? []
    }
}

public struct RespuestaSincronizar: Decodable, Sendable {
    enum CodingKeys: String, CodingKey {
        case elementos, etiquetasDefinidas, servidorEn, masDisponible
    }

    public let elementos: [Elemento]
    public let etiquetasDefinidas: [EtiquetaDefinida]
    public let servidorEn: MarcaDeTiempo
    public let masDisponible: Bool

    public init(from decoder: any Decoder) throws {
        let campos = try decoder.container(keyedBy: CodingKeys.self)
        elementos = try campos.decodeIfPresent([Elemento].self, forKey: .elementos) ?? []
        etiquetasDefinidas =
            try campos.decodeIfPresent([EtiquetaDefinida].self, forKey: .etiquetasDefinidas) ?? []
        servidorEn = try campos.decodeIfPresent(MarcaDeTiempo.self, forKey: .servidorEn) ?? 0
        masDisponible = try campos.decodeIfPresent(Bool.self, forKey: .masDisponible) ?? false
    }
}

/// Un fallo hablando con el servidor, ya traducido a algo que se puede leer
/// en voz alta. `codigo` es el estado HTTP cuando lo hay; sin él, es que no
/// se llegó a hablar con nadie.
public struct ErrorApi: Error, CustomStringConvertible, Sendable {
    public let mensaje: String
    public let codigo: Int?

    public var description: String { mensaje }

    /// El token de acceso ha caducado y toca renovarlo.
    public var esSesionCaducada: Bool { codigo == 401 }
}

/// Cliente HTTP del backend. El contrato completo está en
/// `backend/docs/CONTRATO-API.md`.
///
/// No decide nada: traduce llamadas a peticiones y respuestas a tipos. Quién
/// reintenta y cuándo es cosa de `Sesion`.
public struct ClienteApi: Sendable {
    private let urlBase: String
    private let sesionHttp: URLSession

    public init(urlBase: String, sesionHttp: URLSession = .sesionPorDefecto) {
        self.urlBase = urlBase.replacingOccurrences(
            of: "/+$",
            with: "",
            options: .regularExpression
        )
        self.sesionHttp = sesionHttp
    }

    // MARK: - Autenticación

    /// Dirección con la que arranca el login. No se pide desde aquí: se abre
    /// en la hoja de autenticación del sistema, y el servidor responde con
    /// una redirección al consentimiento del proveedor.
    public func urlIniciarLogin(proveedor: String, estado: String, esquema: String) -> URL? {
        var partes = URLComponents(string: "\(urlBase)/auth/iniciar")
        partes?.queryItems = [
            URLQueryItem(name: "proveedor", value: proveedor),
            URLQueryItem(name: "modo", value: "deeplink"),
            URLQueryItem(name: "estado", value: estado),
            URLQueryItem(name: "esquema", value: esquema),
        ]
        return partes?.url
    }

    /// Cambia el código de canje (un solo uso, unos 60 segundos de vida) por
    /// los tokens de sesión.
    public func canjear(codigoCanje: String) async throws -> RespuestaCanje {
        try await pedir("/auth/canjear", cuerpo: ["codigoCanje": codigoCanje])
    }

    /// Solo sirve contra un servidor con `PERMITIR_LOGIN_DEV=true`. Queda
    /// como la única forma de entrar en el backend de pruebas local, que
    /// arranca con credenciales de Google falsas.
    public func loginDeDesarrollo(email: String) async throws -> RespuestaCanje {
        try await pedir("/auth/dev-login", cuerpo: ["email": email])
    }

    public func renovar(tokenRefresco: String) async throws -> RespuestaCanje {
        try await pedir("/auth/renovar", cuerpo: ["tokenRefresco": tokenRefresco])
    }

    public func cerrarSesion(tokenRefresco: String) async throws {
        let _: RespuestaVacia = try await pedir(
            "/auth/logout",
            cuerpo: ["tokenRefresco": tokenRefresco]
        )
    }

    // MARK: - Metadatos

    public func metadatos(url: String, tokenAcceso: String) async throws -> RespuestaMetadatos {
        try await pedir("/metadatos", cuerpo: ["url": url], tokenAcceso: tokenAcceso)
    }

    // MARK: - Sincronización

    public func bajarCambios(
        desde: MarcaDeTiempo,
        tokenAcceso: String,
        limite: Int = 300
    ) async throws -> RespuestaSincronizar {
        var partes = URLComponents(string: "\(urlBase)/sincronizar")
        partes?.queryItems = [
            URLQueryItem(name: "desde", value: String(desde)),
            URLQueryItem(name: "limite", value: String(limite)),
        ]
        guard let url = partes?.url else {
            throw ErrorApi(mensaje: "la dirección del servidor no es válida", codigo: nil)
        }
        var peticion = URLRequest(url: url)
        peticion.httpMethod = "GET"
        peticion.setValue("Bearer \(tokenAcceso)", forHTTPHeaderField: "Authorization")
        return try await enviar(peticion)
    }

    public func subirCambios(
        elementos: [Elemento],
        etiquetas: [EtiquetaDefinida],
        tokenAcceso: String
    ) async throws -> RespuestaPush {
        struct CuerpoPush: Encodable {
            let elementos: [Elemento]?
            let etiquetasDefinidas: [EtiquetaDefinida]?
        }
        // Los dos son opcionales, pero hace falta al menos uno: se manda solo
        // lo que hay, para no enviar listas vacías que el servidor rechaza.
        let cuerpo = CuerpoPush(
            elementos: elementos.isEmpty ? nil : elementos,
            etiquetasDefinidas: etiquetas.isEmpty ? nil : etiquetas
        )
        return try await pedir("/sincronizar", cuerpoCodificable: cuerpo, tokenAcceso: tokenAcceso)
    }

    // MARK: - Fontanería interna

    private struct RespuestaVacia: Decodable {}

    private func pedir<T: Decodable>(
        _ ruta: String,
        cuerpo: [String: String],
        tokenAcceso: String? = nil
    ) async throws -> T {
        try await pedir(ruta, cuerpoCodificable: cuerpo, tokenAcceso: tokenAcceso)
    }

    private func pedir<T: Decodable, C: Encodable>(
        _ ruta: String,
        cuerpoCodificable: C,
        tokenAcceso: String? = nil
    ) async throws -> T {
        guard let url = URL(string: "\(urlBase)\(ruta)") else {
            throw ErrorApi(mensaje: "la dirección del servidor no es válida", codigo: nil)
        }
        var peticion = URLRequest(url: url)
        peticion.httpMethod = "POST"
        peticion.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let tokenAcceso {
            peticion.setValue("Bearer \(tokenAcceso)", forHTTPHeaderField: "Authorization")
        }
        peticion.httpBody = try JSONEncoder().encode(cuerpoCodificable)
        return try await enviar(peticion)
    }

    private func enviar<T: Decodable>(_ peticion: URLRequest) async throws -> T {
        let datos: Data
        let respuesta: URLResponse
        do {
            (datos, respuesta) = try await sesionHttp.data(for: peticion)
        } catch {
            // Los mensajes de URLSession llegan en inglés y contando cosas de
            // red que aquí no ayudan. Esta frase acaba leída en voz alta.
            // En minúscula y sin verbo: esto se lee solo ("Sin conexión con
            // el servidor") y también como causa detrás de la acción que
            // falló ("No se pudo sincronizar: sin conexión con el servidor").
            throw ErrorApi(mensaje: "sin conexión con el servidor", codigo: nil)
        }

        let codigo = (respuesta as? HTTPURLResponse)?.statusCode ?? 0
        guard (200...299).contains(codigo) else {
            throw ErrorApi(mensaje: Self.mensajeDeError(datos, codigo: codigo), codigo: codigo)
        }
        if datos.isEmpty, let vacia = RespuestaVacia() as? T {
            return vacia
        }
        do {
            return try JSONDecoder().decode(T.self, from: datos)
        } catch {
            throw ErrorApi(mensaje: "el servidor respondió algo que no se entiende", codigo: codigo)
        }
    }

    /// El cuerpo de error del contrato es `{"error": "..."}`.
    private static func mensajeDeError(_ datos: Data, codigo: Int) -> String {
        struct CuerpoError: Decodable {
            let error: String?
        }
        if let cuerpo = try? JSONDecoder().decode(CuerpoError.self, from: datos),
            let detalle = cuerpo.error,
            !detalle.isEmpty
        {
            return detalle
        }
        return "el servidor respondió con un error \(codigo)"
    }
}

extension URLSession {
    /// Diez segundos: lo que esperaba la app de Expo. Más que eso, en un
    /// móvil, es tiempo mirando una pantalla que no dice nada.
    public static let sesionPorDefecto: URLSession = {
        let configuracion = URLSessionConfiguration.default
        configuracion.timeoutIntervalForRequest = 10
        return URLSession(configuration: configuracion)
    }()
}

File: app-android/fontaneria/src/main/kotlin/com/jmortizsilva/guardarenlaces/fontaneria/ClienteApi.kt
package com.jmortizsilva.guardarenlaces.fontaneria

import com.jmortizsilva.guardarenlaces.dominio.Elemento
import com.jmortizsilva.guardarenlaces.dominio.EtiquetaDefinida
import com.jmortizsilva.guardarenlaces.dominio.Login
import com.jmortizsilva.guardarenlaces.dominio.MarcaDeTiempo
import com.jmortizsilva.guardarenlaces.dominio.MetadatosExtraidos
import com.jmortizsilva.guardarenlaces.dominio.Proveedor
import com.jmortizsilva.guardarenlaces.dominio.TipoElemento
import com.jmortizsilva.guardarenlaces.dominio.jsonDelContrato
import java.io.IOException
import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerializationException
import kotlinx.serialization.serializer
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

/** Lo que responde el servidor al canjear un código o al renovar la sesión. */
@Serializable
data class RespuestaCanje(
    val tokenAcceso: String,
    val expiraEn: MarcaDeTiempo,
    val tokenRefresco: String,
    val usuario: UsuarioApi? = null,
)

@Serializable data class UsuarioApi(val id: Long, val email: String, val proveedor: String)

@Serializable
data class RespuestaMetadatos(
    val titulo: String? = null,
    val descripcion: String? = null,
    val imagenUrl: String? = null,
    val tipo: String = "enlace",
) {
    val comoMetadatos
        get() = MetadatosExtraidos(titulo, descripcion, imagenUrl, TipoElemento.desde(tipo))
}

/**
 * Una entrada que el servidor no ha aplicado. Hay que sacarla de la cola igualmente: no se va a
 * aceptar por mucho que se insista, y dejarla dentro reenvía el lote entero en cada sincronización,
 * para siempre.
 */
@Serializable
data class EntradaRechazada(
    val id: String,
    /** `sin_url`, `sin_nombre` o `no_aplicable` (ver CONTRATO-API.md). */
    val motivo: String,
)

@Serializable
data class RespuestaPush(
    val elementos: List<Elemento> = emptyList(),
    val rechazados: List<EntradaRechazada> = emptyList(),
    val etiquetasDefinidas: List<EtiquetaDefinida> = emptyList(),
    val etiquetasRechazadas: List<EntradaRechazada> = emptyList(),
)

@Serializable
data class RespuestaSincronizar(
    val elementos: List<Elemento> = emptyList(),
    val etiquetasDefinidas: List<EtiquetaDefinida> = emptyList(),
    val servidorEn: MarcaDeTiempo = 0,
    val masDisponible: Boolean = false,
)

/**
 * Un fallo hablando con el servidor, ya traducido a algo que se puede leer en voz alta. `codigo` es
 * el estado HTTP cuando lo hay; sin él, es que no se llegó a hablar con nadie.
 */
class ErrorApi(val mensaje: String, val codigo: Int?) : Exception(mensaje) {
    /** El token de acceso ha caducado y toca renovarlo. */
    val esSesionCaducada
        get() = codigo == 401
}

/**
 * Cliente HTTP del backend. El contrato completo está en `backend/docs/CONTRATO-API.md`.
 *
 * No decide nada: traduce llamadas a peticiones y respuestas a tipos. Quién reintenta y cuándo es
 * cosa de `Sesion`.
 */
class ClienteApi(urlBase: String, private val http: OkHttpClient = clienteHttpPorDefecto) {
    private val base: HttpUrl =
        requireNotNull(urlBase.trimEnd('/').toHttpUrlOrNull()) {
            "la dirección del servidor no es válida: $urlBase"
        }

    // Autenticación

    /**
     * Dirección con la que arranca el inicio de sesión. No se pide desde aquí: se abre en el
     * navegador, y el servidor responde con una redirección al consentimiento del proveedor.
     */
    fun urlIniciarLogin(proveedor: Proveedor, estado: String): String =
        ruta("auth/iniciar")
            .newBuilder()
            .addQueryParameter("proveedor", proveedor.valor)
            .addQueryParameter("modo", "deeplink")
            .addQueryParameter("estado", estado)
            .addQueryParameter("esquema", Login.ESQUEMA)
            .build()
            .toString()

    /** Cambia el código de canje (un solo uso, unos 60 segundos de vida) por los tokens. */
    suspend fun canjear(codigoCanje: String): RespuestaCanje =
        pedir("auth/canjear", mapOf("codigoCanje" to codigoCanje))

    /** Solo contra un servidor con `PERMITIR_LOGIN_DEV=true`. */
    suspend fun loginDeDesarrollo(email: String): RespuestaCanje =
        pedir("auth/dev-login", mapOf("email" to email))

    suspend fun renovar(tokenRefresco: String): RespuestaCanje =
        pedir("auth/renovar", mapOf("tokenRefresco" to tokenRefresco))

    suspend fun cerrarSesion(tokenRefresco: String) {
        pedir<Map<String, String>, Nada>("auth/logout", mapOf("tokenRefresco" to tokenRefresco))
    }

    // Metadatos

    suspend fun metadatos(url: String, tokenAcceso: String): RespuestaMetadatos =
        pedir("metadatos", mapOf("url" to url), tokenAcceso)

    // Sincronización

    suspend fun bajarCambios(
        desde: MarcaDeTiempo,
        tokenAcceso: String,
        limite: Int = 300,
    ): RespuestaSincronizar {
        val url =
            ruta("sincronizar")
                .newBuilder()
                .addQueryParameter("desde", desde.toString())
                .addQueryParameter("limite", limite.toString())
                .build()
        val peticion =
            Request.Builder().url(url).get().header("Authorization", "Bearer $tokenAcceso").build()
        return enviar(peticion, serializer())
    }

    @Serializable
    private class CuerpoPush(
        val elementos: List<Elemento>? = null,
        val etiquetasDefinidas: List<EtiquetaDefinida>? = null,
    )

    suspend fun subirCambios(
        elementos: List<Elemento>,
        etiquetas: List<EtiquetaDefinida>,
        tokenAcceso: String,
    ): RespuestaPush =
        // Los dos son opcionales, pero hace falta al menos uno: se manda solo lo que hay, para no
        // enviar listas vacías.
        pedir(
            "sincronizar",
            CuerpoPush(elementos.ifEmpty { null }, etiquetas.ifEmpty { null }),
            tokenAcceso,
        )

    // Por dentro

    /** Para las respuestas que no traen nada que leer, como cerrar sesión. */
    @Serializable private class Nada

    private fun ruta(camino: String): HttpUrl = base.newBuilder().addPathSegments(camino).build()

    private suspend inline fun <reified C, reified T> pedir(
        camino: String,
        cuerpo: C,
        tokenAcceso: String? = null,
    ): T {
        val json = jsonDelContrato.encodeToString(serializer<C>(), cuerpo)
        val peticion =
            Request.Builder()
                .url(ruta(camino))
                .post(json.toRequestBody(tipoJson))
                .apply { if (tokenAcceso != null) header("Authorization", "Bearer $tokenAcceso") }
                .build()
        return enviar(peticion, serializer())
    }

    private suspend fun <T> enviar(peticion: Request, lector: KSerializer<T>): T {
        val (codigo, cuerpo) =
            try {
                http.newCall(peticion).ejecutar()
            } catch (_: IOException) {
                // Los mensajes de OkHttp llegan en inglés y contando cosas de red que aquí no
                // ayudan. Esta frase acaba leída en voz alta, sola («Sin conexión con el
                // servidor») o detrás de la acción que falló («No se pudo sincronizar: sin
                // conexión con el servidor»); por eso va en minúscula y sin verbo.
                throw ErrorApi("sin conexión con el servidor", null)
            }

        if (codigo !in 200..299) throw ErrorApi(mensajeDeError(cuerpo, codigo), codigo)
        return try {
            jsonDelContrato.decodeFromString(lector, cuerpo.ifBlank { "{}" })
        } catch (_: SerializationException) {
            throw ErrorApi("el servidor respondió algo que no se entiende", codigo)
        } catch (_: IllegalArgumentException) {
            throw ErrorApi("el servidor respondió algo que no se entiende", codigo)
        }
    }

    @Serializable private class CuerpoError(val error: String? = null)

    /** El cuerpo de error del contrato es `{"error": "..."}`. */
    private fun mensajeDeError(cuerpo: String, codigo: Int): String {
        val detalle =
            try {
                jsonDelContrato.decodeFromString<CuerpoError>(cuerpo).error
            } catch (_: SerializationException) {
                null
            } catch (_: IllegalArgumentException) {
                null
            }
        return detalle?.ifBlank { null } ?: "el servidor respondió con un error $codigo"
    }

    private companion object {
        val tipoJson = "application/json".toMediaType()
    }
}
